ILM migrate data between tiers

andreidan opened the pull request:
This adds ILM support for automatically migrating the managed
indices between data tiers.

This proposal makes use of a MigrateAction that is injected 
(similar to how the Unfollow action is injected) in phases that
don't define index allocation rules using the AllocateAction or
don't explicitly define the MigrateAction itself (regardless if it's
enabled or disabled).

Relates to #60848

andreidan committed:
Enable the _tier attribute as a node filter

andreidan committed:
ILM: inject a migrate step to migrate data between data tiers

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/AllocationRoutedStep.java:
```diff
@@ -40,8 +43,23 @@
+        ALL_CLUSTER_SETTINGS = allSettings;
+    }
+
+    private static final AllocationDeciders ALLOCATION_DECIDERS = new AllocationDeciders(
+        List.of(
+            new FilterAllocationDecider(Settings.EMPTY, new ClusterSettings(Settings.EMPTY,
+                ClusterSettings.BUILT_IN_CLUSTER_SETTINGS)),
+            new DataTierAllocationDecider(new ClusterSettings(Settings.EMPTY, ALL_CLUSTER_SETTINGS))
+        )
```
should we make this configurable and serializable for the step? (ie. use just the one allocation decider depending on where it is used)
I'd say we want a fully allocated index so verifying both always is alright. What do you think?

dakrone replied:
Hmm... I actually think that we may want to split checking the allocation for the migrate action into a separate step. For example, the allocation routed step currently has a pretty generic message (`Waiting for [n] shards to be allocated to nodes matching the given filters`). I think if we split this into a new `MigrationRouted` step we could give it a much better explanation, for example, something like:

`waiting [23m] for [3] shards to be allocated on nodes with the [data_warm] tier`

additionally, I think we could also even throw an error to signal to the user when things were in a bad state, something like:

`exception waiting for index to be moved to the [data_cold] tier, there are currently no [data_cold] nodes in the cluster`

Then the step could be retryable (so we check every 10 minutes) and it at least gives us a way of signaling to a user (alerting on the ilm-history index for example) when they are in an irreconcilable position and need to adjust their cluster.

What do you think?

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:
```diff
@@ -0,0 +1,112 @@
+    public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
+        if (enabled) {
+            Map<String, String> include = Map.of("_tier", "data_" + phase);
+            AllocateAction migrateDataAction = new AllocateAction(null, include, null, null);
```
should we remove the possible`_require` and `_exclude` settings (which might've been set manually before) to make sure they don't invalidate the "migrate to the next data tier" goal?

dakrone replied:
That's a good question, I'm not sure we'd want to piecemeal remove all of them, because they could be something for an *additional* attribute that we want to preserve between phases.

I will have to think on it a bit, it also makes me wonder whether we should make it configurable whether all other index-level allocation filtering settings are removed/preserved when doing the migration.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:
```diff
@@ -0,0 +1,112 @@
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), ENABLED_FIELD);
+    }
+
+    private final boolean enabled;
```
`enabled` should be true by default I think?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:
```diff
@@ -0,0 +1,112 @@
+        a -> new MigrateAction((boolean) a[0]));
+
+    static {
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), ENABLED_FIELD);
```
I think we could use `optionalConstructorArg` here, so a user could specify `"migrate": {}` if they wanted the default (would would be enabled)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/MigrateAction.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
+        if (enabled) {
+            Map<String, String> include = Map.of("_tier", "data_" + phase);
```
Can you add a sanity assert that the value in this map is always one of the `DataTier.DATA_(HOT|WARM|COLD|FROZEN)` values? (in case those names are changed)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverAction.java:
```diff
@@ -27,7 +27,7 @@
 
 /**
- * A {@link LifecycleAction} which deletes the index.
+ * A {@link LifecycleAction} which rolls over the index.
```
Good catch :D

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -40,18 +43,18 @@
     static final List<String> ORDERED_VALID_HOT_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, RolloverAction.NAME,
```
I think we should add the migrate action explicitly to the hot actions as well (the hot phase doesn't have to start at `min_age: 0ms`)

andreidan replied:
Interesting. Could this potentially cause confusion if new indices are configured to go to the hot nodes (via https://github.com/elastic/elasticsearch/pull/61342 ) and a policy has the migrate action configured to be disabled?

dakrone replied:
It could, but I was thinking of a pre-existing index having an ILM policy applied, it would be a bit strange if only the warm/cold/frozen phases moved the data around, instead of starting it on the `hot` nodes

andreidan replied:
Added a couple of tests for these scenarios in https://github.com/elastic/elasticsearch/pull/61377/commits/fe1fb8f5e70847ccdfb5e52e519aad6c09164e2b

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -86,12 +89,38 @@ public String getWriteableName() {
         return orderedPhases;
     }
 
+    boolean shouldMigrateDataToTiers(Phase phase) {
```
This might be better named `shouldInjectMigrateStepForPhase` I think

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -86,12 +89,38 @@ public String getWriteableName() {
+    boolean shouldMigrateDataToTiers(Phase phase) {
+        AllocateAction allocateAction = (AllocateAction) phase.getActions().get(AllocateAction.NAME);
+        if (allocateAction != null) {
+            if (allocateAction.getExclude().isEmpty() == false ||
+                allocateAction.getInclude().isEmpty() == false ||
+                allocateAction.getRequire().isEmpty() == false) {
```
This could use your `definesAllocationRules(allocateAction)` helper here instead of manually checking, though I think maybe that helper needs to use `isEmpty` rather than checking for `null` so that an allocate action with `"include": {}` doesn't trip the assertion

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -226,5 +255,24 @@ public void validate(Collection<Phase> phases) {
+    }
+
+    private boolean definesAllocationRules(AllocateAction action) {
+        return action.getRequire() != null || action.getInclude() != null || action.getExclude() != null;
```
These should use the `.isEmpty()` check rather than checking for `null`, since they could be `"include": {}` and be non-null but not actually have any allocation rules

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -226,5 +255,24 @@ public void validate(Collection<Phase> phases) {
+        }
+    }
+
+    private boolean definesAllocationRules(AllocateAction action) {
```
```suggestion
    private static boolean definesAllocationRules(AllocateAction action) {
```

dakrone reviewed (commented):
Thanks for working on this Andrei! I left some thoughts on the implementation.

I also was thinking/wondering, is this something we think should have an index level opt-out setting? I'm thinking of our internal indices where we might be using an ILM policy but don't know the topology of the cluster (and honestly don't care about it living on a specific data tier).

I am also wondering if maybe should have the setting be a more generic "opt out of automatic tier management" that would opt out of allocating on hot nodes by default. We could then potentially use this setting for internal indices that we don't want to have to worry about, like `.security` or `.kibana`.

What do you think?

andreidan committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

andreidan committed:
Add a DataTierMigrationRoutedStep and cleanup

andreidan committed:
Fix tests

andreidan committed:
Fix test

andreidan committed:
Remove invalid tier name setting validation from ILM step

We already validate the data tier setting values as part of the
index settings validation.

andreidan committed:
Add test for DataTierMigrationRoutedStep

andreidan committed:
Disable data tier migration in the full policy we use for timeseries ITS

We'll have dedicated data tier integration tests that'll have data migration
enabled.

andreidan committed:
Remove unused import

andreidan committed:
Revert "Disable data tier migration in the full policy we use for timeseries ITS"

This reverts commit 6c05591ce911008dc22e4dd342991f56deb9f4f0.

andreidan committed:
Add MigrateActionTest

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

andreidan committed:
Don't randomly pick a disabled migrate action as it has no steps

andreidan committed:
Add ILM tier migratino IT

andreidan committed:
Use 0 replicas in docs test as `check-migration` waits for replicas to start

andreidan committed:
Fix ILMHistoryTests

andreidan commented:
@elasticmachine update branch

elasticmachine commented:
expected head sha didn’t match current head ref.

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

andreidan committed:
Disable migrate action in test that waits for shard allocations

andreidan committed:
Add ILM client MigrateAction

andreidan committed:
Remove unused import

andreidan committed:
Fix HLRC tests

andreidan committed:
Add tests for conflicting auto allocate to hot nodes and MigrateAction configurations

andreidan committed:
Use valid tier name in test

andreidan commented on client/rest-high-level/src/main/java/org/elasticsearch/client/ilm/LifecyclePolicy.java:
```diff
@@ -57,10 +57,13 @@
+        ALLOWED_ACTIONS.put("hot", Sets.newHashSet(UnfollowAction.NAME, SetPriorityAction.NAME, MigrateAction.NAME, RolloverAction.NAME));
+        ALLOWED_ACTIONS.put("warm", Sets.newHashSet(UnfollowAction.NAME, SetPriorityAction.NAME, MigrateAction.NAME, AllocateAction.NAME,
+            ForceMergeAction.NAME, ReadOnlyAction.NAME, ShrinkAction.NAME));
+        ALLOWED_ACTIONS.put("cold", Sets.newHashSet(UnfollowAction.NAME, SetPriorityAction.NAME, MigrateAction.NAME, AllocateAction.NAME,
+            FreezeAction.NAME));
+        ALLOWED_ACTIONS.put("frozen", Sets.newHashSet(UnfollowAction.NAME, SetPriorityAction.NAME, MigrateAction.NAME, AllocateAction.NAME,
+            FreezeAction.NAME));
```
We need to add client support for the SearchableSnapshot action too

andreidan committed:
Implement hashcode

andreidan commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

andreidan committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

andreidan committed:
Make method static

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/ILM+SLM)

andreidan commented:
@elasticmachine run elasticsearch-ci/packaging-sample-windows

andreidan commented:
> I am also wondering if maybe should have the setting be a more generic "opt out of automatic tier management" that would opt out of allocating on hot nodes by default
> 
> @dakrone I think this makes sense

I've started working on this in a separate PR (to avoid this rather large PR growing even more)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DataTierMigrationRoutedStep.java:
```diff
@@ -0,0 +1,118 @@
+            String tier = INDEX_ROUTING_INCLUDE_SETTING.get(idxMeta.getSettings());
+            boolean targetTierNodeFound = false;
+            for (DiscoveryNode node : clusterState.nodes()) {
+                for (DiscoveryNodeRole role : node.getRoles()) {
+                    if (role.roleName().equals(tier)) {
```
I don't think this will work, because it's checking for the exact role, so for example, if the allocation were set to `data_warm`, and we had `data` nodes but not nodes with the exact `data_warm` role, this would say that they were not found (even though they are).

We should check for either an exact match or the presence of any nodes with the `data` role

andreidan replied:
Great point

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DataTierMigrationRoutedStep.java:
```diff
@@ -0,0 +1,118 @@
+                    "tier" + (targetTierNodeFound ? "" : " but there are currently no [%s] nodes in the cluster"),
+                index, getKey().getAction(), allocationPendingAllShards, tier, tier);
+            logger.debug(statusMessage);
+            return new Result(false, new AllocationInfo(idxMeta.getNumberOfReplicas(), allocationPendingAllShards, true, statusMessage));
```
I think we discussed making this throw an exception if there were no nodes with the target tier in the cluster, so that the `ERROR` state was signaled (and the user could alert for it), but this only sets the result to `false`.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -37,21 +40,21 @@
     static final List<String> VALID_PHASES = Arrays.asList(HOT_PHASE, WARM_PHASE, COLD_PHASE, FROZEN_PHASE, DELETE_PHASE);
-    static final List<String> ORDERED_VALID_HOT_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, RolloverAction.NAME,
-        ForceMergeAction.NAME);
+    static final List<String> ORDERED_VALID_HOT_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, MigrateAction.NAME,
```
We should remove this from the hot phase (I know you know this already, this is just a placeholder comment)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/step/info/AllocationInfo.java:
```diff
@@ -0,0 +1,114 @@
+import java.io.IOException;
+import java.util.Objects;
+
+public class AllocationInfo implements ToXContentObject {
```
Could you add javadocs please?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/step/info/AllocationInfo.java:
```diff
@@ -0,0 +1,114 @@
+    private final boolean allShardsActive;
+    private final String message;
+
+    static final ParseField ACTUAL_REPLICAS = new ParseField("actual_replicas");
```
I think "actual_replicas" is a little difficult to tell what it means, maybe there's a clearer name we could use?

andreidan replied:
Yes, I agree. It'll be a breaking change, but possibly going with the well known `number_of_replicas` would be better?

dakrone commented on x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunnerTests.java:
```diff
@@ -91,6 +92,8 @@
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.when;
 
+;
+
```
Whoops I think this is a typo that can be removed?

dakrone reviewed (changes requested):
Thanks for working on this Andrei, I left some more comments.

andreidan committed:
Check for "data" role in `check-migration` step

andreidan committed:
Document AllocationInfo and rename `actual_replicas` to `number_of_replicas`

andreidan committed:
Disallow the migrate action in the hot phase

andreidan committed:
Remove rogue ;

andreidan committed:
Fix HLRC tests

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

andreidan committed:
Fix ILMHistoryTests

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -40,18 +43,18 @@
-        FreezeAction.NAME, SearchableSnapshotAction.NAME);
-    static final List<String> ORDERED_VALID_FROZEN_ACTIONS = Arrays.asList(SetPriorityAction.NAME, UnfollowAction.NAME, AllocateAction.NAME,
-        FreezeAction.NAME, SearchableSnapshotAction.NAME);
+        MigrateAction.NAME, AllocateAction.NAME, ShrinkAction.NAME, ForceMergeAction.NAME);
```
I realized something about this, I think we should do `AllocateAction` *prior* to the migrate action, or else we are liable to get "stuck".

For example, if a user had multiple hot nodes and only a single warm node, we would try to migrate an index with one replica from hot to warm, but the replica could never allocate, since there is only a single node. Instead, we should do the `allocate` action first (where a user could set `number_of_replicas` to 0) before migrating to the next tier.

What do you think?

dakrone replied:
I also realize we can get stuck the opposite way too, so either way it's possible to get stuck, but I think that doing allocate first is still the way we should go. Curious about what you think

andreidan replied:
You make a great point Lee. I think we should go forward this way as it also keeps the "number of replicas is modified before the allocations change" behaviour consistent with how the `allocate` action works.

Maybe it's a bit more of a future talk, but I do wonder if it'd be a good time to extract the "change number of replicas" into its own action. I think being conflated with the allocation rules in the `allocate` action is generally confusing, but even more so now where we'd expect the users to use both the `allocate` and the `migrate` (admittedly, not necessarily explicitly declaring it) actions to achieve the "reduce number of replicas and then relocate index" scenario.
I think a `change-replicas-number` action would be welcome at this stage.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DataTierMigrationRoutedStep.java:
```diff
@@ -0,0 +1,119 @@
+            return new Result(false, null);
+        }
+        if (ActiveShardCount.ALL.enoughShardsActive(clusterState, index.getName()) == false) {
+            logger.debug("[{}] lifecycle action for index [{}] cannot make progress because not all shards are active",
+                    getKey().getAction(), index.getName());
```
I think in this one we can be more specific in the debug message, so perhaps:

```
[check-migration] migration for index [foo] to the [data_cold] tier cannot progress, as not all shards are active
```

(we can pull the tier name from the index setting)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DataTierMigrationRoutedStep.java:
```diff
@@ -0,0 +1,119 @@
+            logger.debug(statusMessage);
+            return new Result(false, new AllocationInfo(idxMeta.getNumberOfReplicas(), allocationPendingAllShards, true, statusMessage));
+        } else {
+            logger.debug("{} lifecycle action for [{}] complete", index, getKey().getAction());
```
same here about log messages, perhaps:

```
[check-migration] migration of index [foo] to tier [data_cold] complete
```

It's definitely a lot nicer to grep for in logs :)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/step/info/AllocationInfo.java:
```diff
@@ -0,0 +1,120 @@
+     * Builds the AllocationInfo representing a cluster state with a routing table that has all the shards active for a particular index
+     * but there are still {@link #numberShardsLeftToAllocate} left to be allocated.
+     */
+    public static AllocationInfo allShardsActiveAllocationInfo(long actualReplicas, long numberShardsLeftToAllocate) {
```
Super minor, but it's confusing to call the parameter here (and in the function above) "`actualReplicas`", it makes it seem like there should be a "`allocatedReplicas`" parameter somewhere as well. Maybe we can just call in `numReplicas`?

andreidan replied:
Totally, my grep skills failed me.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:
```diff
@@ -226,5 +250,24 @@ public void validate(Collection<Phase> phases) {
+            .collect(Collectors.joining(","));
+        if (Strings.hasText(phasesWithConflictingMigrationActions)) {
+            throw new IllegalArgumentException("phases [" + phasesWithConflictingMigrationActions + "] specify an enabled " +
+                MigrateAction.NAME + " action and the " + AllocateAction.NAME + " action. specify only one data migration action in these" +
```
I think rather than:

```
phases [warm,cold] specify an enabled migrate action and the allocate action. specify only one data migration action in these phases
```

We should clarify that it's the actual allocation rules that cause problems (they're free to adjust replica count as much as they'd like):

```
phase [warm,cold] specifies an enabled migrate action and an allocate action with allocation rules, specify only a single data migration in each phase
```

dakrone commented on x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/DataTiersMigrationsTests.java:
```diff
@@ -0,0 +1,149 @@
+        });
+
+        logger.info("starting cold data node");
+        internalCluster().startNode(coldNode(Settings.EMPTY));
```
Should we have the check for "cold" node allocation before starting up the frozen node? We have it for the other phases, so probably best to check for it just so it doesn't end up disappearing for only the cold phase some time in the future

dakrone reviewed (changes requested):
Awesome, this is looking really close, I left one comment about the execution order (I think 'allocate' should come before 'migrate') and a few other really really minor quibbilings, but nothing major

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

andreidan committed:
Remove frozen phase from HLRC

andreidan committed:
Execute the allocate action before migrate

andreidan committed:
Adjust test to wait until ILM completes in the cold phase

andreidan committed:
Log and exception messages

andreidan committed:
Merge branch 'master' into ilm-migrate-data-between-tiers

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DataTierMigrationRoutedStep.java:
```diff
@@ -0,0 +1,119 @@
+
+    @Override
+    public int hashCode() {
+        return 711;
```
I think this needs to incorporate `super.hashCode()` or else the hash code will be the same regardless of stepkeys

dakrone reviewed (approved):
LGTM, thanks for iterating on this Andrei! I left one more minor comment, but it's pretty small

andreidan committed:
Drop equals and hashcode (in favour of the ones in Step)
